designs/discontinue_usage_of_kube_rbac_proxy.md

```diff
@@ -0,0 +1,285 @@
+in the default scaffold due to the evolving k8s infra, community feedback. Key considerations include the transition to a shared infrastructure requiring
+all images to be published on [registry.k8s.io][registry.k8s.io], the deprecation
+of Google Cloud Platform's [Container Registry](https://cloud.google.com/artifact-registry/docs/transition/prepare-gcr-shutdown), and the fact
+that [kube-rbac-proxy][kube-rbac-proxy] is not yet part of the Kubernetes ecosystem umbrella.
+
+The dependency on a potentially discontinuable Google infrastructure,
+**which is out of our control**, paired with the challenges of maintaining,
+building, or promoting [kube-rbac-proxy][kube-rbac-proxy] images,
+calls for a change.
+
+In this document is proposed replacing the [kube-rbac-proxy][kube-rbac-proxy] within
+[Network Policies][k8s-doc-networkpolicies] follow-up for potentially enhancements
+to protect the metrics endpoint combined with [cert-manager][cert-manager] and a new
+feature introduced in controller-runtime, see [here][cr-pr].
+
+**For the future (when kube-rbac-proxy be part of the k8s umbrella)**, it is proposed the usage of the
+[Plugins API provided by Kubebuilder](./../docs/book/src/plugins/plugins.md),
+to create an [external plugin](./../docs/book/src/plugins/creating-plugins.md)
+to properly integrate the solution with Kubebuilder and provide a helper to allow users to opt-in as please them.
+
+## Open Questions
+
+1. [Network Policies][k8s-doc-networkpolicies] is implemented by the cluster’s CNI. Are we confident that the proposed policy is supported by all the major CNIs in use?
+
+> Besides [Network Policies][k8s-doc-networkpolicies] being part of the core Kubernetes API, their enforcement relies on the CNI plugin installed in
+the Kubernetes cluster. While support and implementation details can vary among CNIs, it seems that the most commonly used ones,
+such as Calico, Cilium, WeaveNet, and Canal, shows to offer  support for NetworkPolicies.
+>
+>Also, there was concern in the past because AWS did not support it. However, this changed,
+>as detailed in their announcement: [Amazon VPC CNI now supports Kubernetes Network Policies](https://aws.amazon.com/blogs/containers/amazon-vpc-cni-now-supports-kubernetes-network-policies/).
+>
+>Moreover, under this proposal users still able to disable/enable this option as please them.
+
+1. NetworkPolicy is a simple firewall and does not provide authn/authz and encryption.
+
+> Yes, that's correct. NetworkPolicy acts as a basic firewall for pods within a Kubernetes cluster, controlling traffic
+> flow at the IP address or port level. However, it doesn't handle authentication (authn), authorization (authz),
+> or encryption directly like kube-rbac-proxy solution.
+>
+> However, if we be able to combine the cert-manager and the new feature provided
+> by controller-runtime we can achieve the same or a superior level of protection
+> without relay in any extra third-party dependency.
```

Comment:
link to where this is documented?

Reply:
Hi @joejulian,

Thank you for your review and time.
Could you please let us know what documentation are you looking for here?
What exactly clarifications do you need in the OpenQuestion answered.

Reply:
Someone who was previously relying on kube-rbac-proxy wants to now follow the new recommended method of using cert-manager and the new mtls feature (if I'm understanding this correctly), where do they find the documentation for that? Could we link that here?

Reply:
That is the proposal for change 
We are proposing it all get done in phases
So, the first phase does not use cert-manager 
See: https://github.com/kubernetes-sigs/kubebuilder/pull/3853

Then, a follow up of this PR we need to add a config in the kustomize
that will allow we patch the serviceMonitor and certs such as we do now for webhooks 
See that we have in the config/default/ a patch for cert-manager + webhooks: 
See https://github.com/kubernetes-sigs/kubebuilder/blob/bd0876b8132ff66da12d8d8a0fdc701fde00f54b/docs/book/src/component-config-tutorial/testdata/project/config/default/kustomization.yaml#L51-L149 and https://github.com/kubernetes-sigs/kubebuilder/blob/master/testdata/project-v4/config/default/manager_webhook_patch.yaml

In the follow up we will need to do something similar 
We will need to have something like

```
# [METRICS WITH HTTPS] To enable the ServiceMonitor using HTTPS uncomment the following line
# Note that for that works you also need ensure that you enable cert-manager in your project
- path: metrics_https_patch.yaml
```

That should patch https://github.com/kubernetes-sigs/kubebuilder/pull/3853/files#diff-8545904ae77cfcef7da403b4fe25b822439217abe9307d5d31b29690140d0fc6

As part of the task we will need also to update the docs to clarify how that works and etc.

Reply:
Hi @joejulian 

It was also added in the proposal.
See that we are adding examples either how the YAML would end up.

Thank you for raise it out.